arch/cortex-m4/src/mpu.rs

```diff
@@ -116,160 +127,496 @@ impl MPU {
-        }
+        regs.ctrl
+            .write(Control::ENABLE::SET + Control::HFNMIENA::CLEAR + Control::PRIVDEFENA::SET);
     }
 
     fn disable_mpu(&self) {
         let regs = &*self.0;
-        regs.control.set(0b0);
+        regs.ctrl.write(Control::ENABLE::CLEAR);
     }
 
-    fn create_region(
-        &self,
-        region_num: usize,
-        start: usize,
-        len: usize,
-        execute: kernel::mpu::ExecutePermission,
-        access: kernel::mpu::AccessPermission,
-    ) -> Option<Region> {
-        if region_num >= 8 {
-            // There are only 8 (0-indexed) regions available
-            return None;
-        }
+    fn number_total_regions(&self) -> usize {
+        let regs = &*self.0;
+        regs.mpu_type.read(Type::DREGION) as usize
+    }
 
-        // There are two possibilities we support:
-        //
-        // 1. The base address is aligned exactly to the size of the region,
-        //    which uses an MPU region with the exact base address and size of
-        //    the memory region.
-        //
-        // 2. Otherwise, we can use a larger MPU region and expose only MPU
-        //    subregions, as long as the memory region's base address is aligned
-        //    to 1/8th of a larger region size.
-
-        if start % len == 0 {
-            // Memory base aligned to memory size - straight forward case
-            let region_len = PowerOfTwo::floor(len as u32);
-            if region_len.exp::<u32>() < 5 {
-                // Region sizes must be 32 Bytes or larger
-                return None;
-            } else if region_len.exp::<u32>() > 32 {
-                // Region sizes must be 4GB or smaller
+    fn allocate_region(
+        &self,
+        unallocated_memory_start: *const u8,
+        unallocated_memory_size: usize,
+        min_region_size: usize,
+        permissions: mpu::Permissions,
+        config: &mut Self::MpuConfig,
+    ) -> Option<mpu::Region> {
+        // Check that no previously allocated regions overlap the unallocated memory.
+        for region in config.regions.iter() {
+            if region.overlaps(unallocated_memory_start, unallocated_memory_size) {
                 return None;
             }
+        }
 
-            let xn = execute as u32;
-            let ap = access as u32;
-            Some(unsafe {
-                Region::new(
-                    (start | 1 << 4 | (region_num & 0xf)) as u32,
-                    1 | (region_len.exp::<u32>() - 1) << 1 | ap << 24 | xn << 28,
-                )
-            })
-        } else {
-            // Memory base not aligned to memory size
+        let region_num = match config.unused_region_number() {
+            Some(number) => number,
+            None => return None,
+        };
+
+        // Logical region
+        let mut start = unallocated_memory_start as usize;
+        let mut size = min_region_size;
+
+        // Region start always has to align to 32 bytes
+        if start % 32 != 0 {
```

Comment:
This works fine but basically, this means that all addresses are divisible by 4 and the two LSB must be zero.
Could be optimized by shifting left by two instead .....but the compiler is probably doing it already!